controllers/userController.js

```diff
@@ -38,6 +38,7 @@ const updateUser = async (req, res) => {
   }
 
   const user = await User.findOne({ _id: req.user.userId }); //get user
+  // Will this set email/name to undefined if tehy weren't given?
   user.email = email; //update properties manually 
   user.name = name;
```

Comment:
Let's say I make a request like the following, where I just want to update my user's name; but not anything else:

![](https://github.com/NatachaKey/ecommerce/assets/9094098/ff3d2e69-be63-4df9-9fd4-f215f0900c99)

Currently, I will get the following response:

![](https://github.com/NatachaKey/ecommerce/assets/9094098/9245b372-5bf1-4bd9-93dd-3a00639681bb)

This is because when I make a request with only `"name"` defined, the destructuring code on line 35 will result in the variable `email` being equal to `undefined`. Then on line 41 the code would set the mongo object `user.email` to undefined. This would then result in a validation error when the user is saved on line 45, which is why we get that message in the response.

In this case, if the user didn't put an email in the request, I think we can assume they just wanted to keep their email as is. And same if they only pass in an email, that probably means they want to update their email but keep their name the same. 

So we should update the code on lines 42 and 43, to use the OR operator `||` to accomplish this:

```
// Update email if request included a non-null value, otherwise keep the email as-is
user.email = email || user.email

// Update name if request included a non-null value, otherwise keep the name as-is
user.name = name || user.name

// Another way we could write these:
if (email) {
  user.email = email
}

if (name) {
  user.name = name
}
```

Reply:
@akosasante thank you! got it! wanted to ask if we can use here findOneAndUpdate() method instead findOne and save()?

Reply:
@akosasante tried to write checkpermissions as a middleware- could you please take a look if it's ok. I added it to updateUser and updateUserPassword- is that right? I mean we already authenticate user, is this new middleware checkPermissions adding extra security  to these http requests? :
 router.route('/updateUser').patch(authenticateUser, checkPermissions, updateUser);
router.route('/updateUserPassword').patch(authenticateUser, checkPermissions, updateUserPassword);

Reply:
and also to review -update and delete routes: router
  .route('/:id')
  .get(getSingleReview)
  .patch(authenticateUser, checkPermissions, updateReview)
  .delete(authenticateUser, checkPermissions, deleteReview);

Reply:
propably if my changes are ok i should add it to order routes too

Reply:
> wanted to ask if we can use here findOneAndUpdate() method instead findOne and save()?

Yes, I believe that would work as well. From what I see in the Mongoose [docs](https://mongoosejs.com/docs/tutorials/findoneandupdate.html) they generally recommend using `save()` instead, but I think they both work in this case.

Reply:
> tried to write checkpermissions as a middleware

Sorry! I thought I removed that review comment :facepalm:  I realized after writing it that actually it'll be difficult to write `checkPermissions` as a middleware, since it needs as one it's inputs, the user from the object that you fetch from the database. But that fetching happens in the controller, and moving that to a middleware would be difficult since you'd need to make it work for all the different types of schemas. So I think what you had is actually the best way, sorry about the confusion!

Reply:
I am not sure if I am passing the correct value so the middleware could compareif this is a user that created order/review/user 
![image](https://github.com/NatachaKey/ecommerce/assets/103755090/d0737aa9-ef1a-4ffb-a383-c0066c93a3a1)

Reply:
@akosasante oh don't worry i will remove that middleware,  i was thinking that in diferent cases we were passing into different parameters (id of the user who created a review/ order ), but if the paramenters would have been always the same it could be a middleware

Reply:
@akosasante I've tried to pass wome wrong arguments in routes documents- when calling this middleware lol but it didn't work, maybe during our Practicum we will cover that

Reply:
> > wanted to ask if we can use here findOneAndUpdate() method instead findOne and save()?
> 
> Yes, I believe that would work as well. From what I see in the Mongoose [docs](https://mongoosejs.com/docs/tutorials/findoneandupdate.html) they generally recommend using `save()` instead, but I think they both work in this case.

@akosasante could you please take a look if I implemented this method correctly (just to be sure)- it works in postman but you never know lol https://github.com/NatachaKey/ecommerce/pull/4
![image](https://github.com/NatachaKey/ecommerce/assets/103755090/14cf2f65-5bef-499c-a2df-094cd67a0309)